### PR TITLE
Phat contract budget adjustment

### PR DESCRIPTION
**Pull request by Index0011**
**Background**
Added some new features to support phat contract to adjust the individual `budget_per_block` for both in Phala and Khala.
**Features**
1.Added a storage value `StandardBudgetPerBlock` which is only to provide a record for contract and has no entry to modify.
2.Added a storage value `BudgetUpdateNonce` to record the latest valid budget update index.
3.Added a storage value `LastBugdetUpdateBlock` to record the latest valid budget update block height.
4.Added a new extrinsic  `set_budget_per_block` to let assigned accounts update `budget_per_block`. Beyond the nonce and block height validation check, it will also forbid any budget larger than the `StandardBudgetPerBlock` to avoid subsidy drains out disaster.
 5.Removed subsidy halving logics in `on_finalized` which will be migrated to the contract.

### CODE REVIEW

File: standalone/runtime/src/lib.rs [hunk 1]
```diff
@@ -1334,6 +1335,17 @@ impl pallet_mq::Config for Runtime {
     type QueueNotifyConfig = msg_routing::MessageRouteConfig;
     type CallMatcher = MqCallMatcher;
 }
+
+pub struct SetBudgetMembers;
```
Comment (h4x3rotab) [inline standalone/runtime/src/lib.rs:1339]: It's ok to use `EnsureSignedBy` here. An easier way is to create a storage item in the pallet to store the account, check the caller in your function by comparing the origin and the stored account, and allow the CouncilOrRoot to edit the account.

**Review by h4x3rotab — COMMENTED** [review]
Let's add @kingsleydon to the reviewer and let him to approve as well.

File: pallets/phala/src/compute/computation.rs [hunk 2]
```diff
@@ -378,6 +388,10 @@ pub mod pallet {
 			session: T::AccountId,
 			p_instant: u32,
 		},
+		BudgetUpdated {
```
[the remark this answers is not in the document]
Comment (kingsleydon) [reply pallets/phala/src/compute/computation.rs:391]: Consider adding an optional payload field to include supplementary information.

File: standalone/runtime/src/lib.rs (hunk 1 above)
Comment (Index0011) [reply standalone/runtime/src/lib.rs:1339]: When i check through the substrate's frame update, i found they prefer use EnsureSignedBy to check validation of callers, can see it in the latest concil pallet, we can implement a `EnsureSignedBy` trait to specify which accounts can change the concil members.

File: pallets/phala/src/compute/computation.rs [hunk 3]
```diff
@@ -525,30 +542,53 @@ pub mod pallet {
...
+			ensure!(
+				nonce > BudgetUpdateNonce::<T>::get(),
+				 Error::<T>::NonceIndexInvalid,
+			);
+			ensure!(
+				block_number > LastBugdetUpdateBlock::<T>::get(),
+				Error::<T>::BudgetUpdateBlockInvalid,
+			);
+			ensure!(
+				budget <= StandardBudgetPerBlock::<T>::get(),
```
[the remark this answers is not in the document]
Comment (Index0011) [reply pallets/phala/src/compute/computation.rs:565]: The Constant value `StandardBudgetPerBlock` is also the current budget for khala-chain, which is the standard value of the total budget that the phat contract can refer.

File: pallets/phala/src/compute/computation.rs [hunk 4]
```diff
@@ -525,30 +542,53 @@ pub mod pallet {
...
+				Error::<T>::BudgetExceedMaxLimit,
+			);
+			BudgetUpdateNonce::<T>::put(nonce);
+			LastBugdetUpdateBlock::<T>::put(block_number);
+			let mut tokenomic = TokenomicParameters::<T>::get().ok_or(Error::<T>::InternalErrorBadTokenomicParameters)?;
```
[the remark this answers is not in the document]
Comment (Index0011) [reply pallets/phala/src/compute/computation.rs:570]: Yes, but i think additional logics may be added and `transactional` is more safe.

File: pallets/phala/src/compute/computation.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Index0011) [reply pallets/phala/src/compute/computation.rs:391]: As we discussed in the meeting, supplementary information will be recorded by phat contract log, no need to record it on chain.

[the remark this answers is not in the document]
Comment (h4x3rotab) [reply pallets/phala/src/compute/computation.rs:391]: Contract Log will not be that easy to access

**Pull request merged**